### PR TITLE
#2956. Add more type void tests. Update existing ones

### PR DESCRIPTION
**Pull request by sgrekhov**

### CODE REVIEW

**Review by eernstg — COMMENTED** [review]
Looks good! I added a few comments, because I think some tests don't test exactly what they claim to test, and also some comments because there are some "intermediate variables" of type `void` that actually reduce the coverage because "it's the same test about variables" in several cases.

**Review by sgrekhov — COMMENTED** [review]
Updated. Some tests where reverted and will be moved to `return` statement tests folder in a separate PR. PTAL.

File: Language/Types/Type_Void/permitted_use_A01_t01.dart [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+/// @assertion In support of the notion that the value of an expression with
+/// static type `void` should be discarded, such objects can only be used in
+/// specific situations: The occurrence of an expression of type `void` is a
+/// compile-time error unless it is permitted according to one of the following
+/// rules.
+/// - In an `⟨expressionStatement⟩` `e;`, `e` may have type void.
+///
+/// @description Checks that it is not an error if in an `⟨expressionStatement⟩`
+/// of the form `e;`, the static type of `e` is `void`.
```
[the remark this answers is not in the document]
Comment (eernstg) [reply Language/Types/Type_Void/permitted_use_A01_t01.dart:13]: Thanks!

File: Language/Types/Type_Void/permitted_use_A04_t02.dart [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+/// @assertion In support of the notion that the value of an expression with
+/// static type `void` should be discarded, such objects can only be used in
+/// specific situations: The occurrence of an expression of type `void` is a
+/// compile-time error unless it is permitted according to one of the following
+/// rules.
+/// ...
+/// - In a parenthesized expression `(e)`, `e` may have type `void`.
+///
+/// @description Checks that in a parenthesized expression `(e)`, `e` may have
+/// type `void` but it is a compile-time error to use it.
```
[the remark this answers is not in the document]
Comment (eernstg) [reply Language/Types/Type_Void/permitted_use_A04_t02.dart]: Thanks!

[the remark this answers is not in the document]
Comment (eernstg) [reply Language/Types/Type_Void/permitted_use_A04_t02.dart]: (Changed 'a' to 'the' in a couple of places.)

File: Language/Types/Type_Void/permitted_use_A05_t02.dart [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+/// @description Checks that if in a conditional expression `e ? e1 : e2`, `e1`
+/// or `e2` have type `void` then the static type of the whole expression is
+/// `void` and it is a compile-time error to use it.
+/// @author sgrekhov22@gmail.com
+
+void foo() => 42;
+
+main() {
+  var e1 = (2 > 1) ? 1 : foo();
+  print(e1);
```
[the remark this answers is not in the document]
Comment (eernstg) [reply Language/Types/Type_Void/permitted_use_A05_t02.dart]: Thanks!

File: Language/Types/Type_Void/permitted_use_A07_t04.dart [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+/// - In a return statement `return e;`, `e` may have type `void` in a number of
+///   situations.
+///
+/// @description Checks that returning a non-null value with non-dynamic static
+/// type from within a `void` method produces compile error.
+/// @author rodionov
+
+void foo() {
+  String s = "foo";
+  return s;
```
[the remark this answers is not in the document]
Comment (eernstg) [reply Language/Types/Type_Void/permitted_use_A07_t04.dart]: Very good, thanks!

**Review by eernstg — APPROVED** [review]
LGTM

**Pull request merged**